### ISSUE TITLE
window.get_active…()—provide unique window ID to accurately target specifc window via resize_move()

### ISSUE THREAD

**Issue by mmhere** [autokey/autokey#306]
## Classification:

Feature (New) / Enhancement

## Reproducibility:

Always

## Version

AutoKey version: 0.95.7

Used GUI (Gtk, Qt, or both): GTK (likely same in Qt)

Installed via:
dpkg --install autokey-common_0.95.7-0_all.deb autokey-gtk_0.95.7-0_all.deb

Linux Distribution:  Ubuntu 16.04

## Summary

window.get_active…()—provide unique window ID to accurately target specifc window via resize_move()

### API tweak request

This script API method...

    thisWin = window.get_active_title() # returns string

...returns only a window title as a string.

This makes it impossible to control specific window(s) with this...

    window.resize_move(thisWin, ...) # assumes thisWin is a string

...when window titles are not unique across the set of all open windows.

As wmctrl is used as the underlying method of window control (according to docs), and that tool accepts a specific window ID (hexadecimal or integer), it would be very nice to access said number that is unique to each window.  Example:

    wmctrl -i -a 0x0a000003
    # alternate:
    wmctrl -i -a 167772163

In autokey Python scripts, you really need to be able to do this:

    thisWinID = window.get_active_id() # give me a unique number!
    window.resize_move(thisWinID, ...) # then I can move that specific window!

Why?

In my situation, there are *usually* multiple windows with the same title!  My style is to lean on the X11 window manager to manage many unique windows, as opposed to leaning on lots of tabs cluttered within windows (I know this fad is popular right now).  In an application such as the Vivaldi browser, I often have several windows with the same title "Vivaldi".

When window.get_active_title() returns "Vivaldi", but there are, say, three such windows, you may not actually control the *active* (top) window with a subsequent call on window.resize(...).  You get whichever one wmctrl decides to select first out of the group of windows with identical titles.

### Suggested Tags

As I do not have write access, I am unable to apply Labels to this issue.  I suggest these:

- enhancement
- Scripting

**Comment by josephj11** [comment]
While this is a valid request, as a workaround, it's not particularly difficult to call `xdotool` or `wmctrl` directly from your script to access any capabilities they provide which AutoKey does not.

`xdotool` returns stacks of window IDs matching various search conditions. I've played with it a bit, but really don't understand it very well. Restricting it to only find visible windows helps a bit.

Keep in mind that AutoKey scripts are just Python3, so, in most cases, if you can figure out how to do something directly in Python, you can use virtually the same code in an AutoKey script.

(+1 for referencing Vivaldi. ;) I'm a fan. )

**Comment by mmhere** [comment]
Good point w.r.t. xdotool.  I hadn't thought of that initially.  I did play around with wmctrl on its own, but it doesn't really give you an easy way (AFAIK) to get the *active* window.  I have played with xdotool in the past but it just didn't come to mind.

So from a shell (or calling out from python) this gets you the window id as an int:

    xdotool getactivewindow # dead simples!

and then of course you could capture the integer id and pass it over to wmctrl for movement.

xdotool started life as a way to automate X11 testing of user actions, so tends to be pretty mouse heavy, as in "move the mouse over here;  press button 1;  move here;  release button 1" but xdotool can retrieve the id.

Still I suppose it would be nice to be able to get things like window IDs directly in autokey to keep it a bit cleaner and do everything "in house."

**Comment by josephj11** [comment]
Just to save you some time:
When you do a search with `xdotool`, it returns a stack of all the window IDs that match that search. X has a different idea of what a window is than humans do. I have seen it return 5 or even more IDs when there was only one matching user window on any desktop (or even none). Using `--onlyvisible` helps with this.
The point is that it can take more than a simple call to do things because the situation as X 
sees it can be a lot more complicated than what we see as users.

`xdotool` has a cool (although challenging) feature called command chaining. If you string together several commands in one `xdotool` invocation it will pass the results from one step to the next. So, the first step could find the window ID and pass it to the next step that acts on that window.

The problem is that all of this happens internally to `xdotool` which makes debugging it kind of like regexes. It either works or it doesn't and your only real option is to stare at it more and hope you'll just see what's wrong. ;)

**Comment by mmhere** [comment]
Thanks for the xdotool notes.  If you do some sort of query, yes, I can see how it would give back a list and some of those are "windows" in the strictest sense, but aren't really what a user considers a window.  (Like maybe some hidden dialog, or a autohide panel or somesuch.)

If, however one does this:

    # I'm being lazy and writing one-liner bashisms here
    targetWinID=$(xdotool getactivewindow)

I think that it always gives you back the top/visible window.

So **IF** the autokey script's goal is to move the top window (for most people including myself this is the "active" window in the sense that it has focus—although certainly some modes of operating window managers can have a window focused that is not the one on 'top") **THEN** the ID returned is the one that can be passed to downstream tools to operate on that window.

You can then follow up with stuff like:

    # assuming the ID is in $targetWinID, again lazy bash code
    wmctrl -i -e "0,1400,200,-1,-1" -r $targetWinID

Of course if you want to move the window relative to its current position you'd have to pick up existing (X,Y) first, calculate new ones then use -e "0,$Xdelta,$Ydelta,-1,-1".

Completeness exercise left to the user.

**Comment by mmhere** [comment]
After further consideration maybe it is possible to do it all in xdotool.

*(Yet in the next comment I present some script that takes a multi-step approach because you can't just feed relative XY values to xdotool;  it does not consider window frame/decoration real estate so a delta given of say 100 pixels doesn't really manifest as 100 pixels in the user's world.)*

Ultimately what I'm trying to achieve is to assign arrow key + modifiers so you can move a window left/right/up/down a certain distance without touching the mouse.  Repeat the keys to move further.

A oneliner using xdotool could take the active window (usually same as "top") and move it down by a relative amount in the Y axis.  I guess this is simply a concrete example of using the command chaining you mentioned.  That's pretty cool.  Thanks for that!  To wit:

    xdotool getactivewindow windowmove --relative x 100

(xdotool should leave x position alone if given the literal argument 'x').  *(only it doesn't really, considering the effect of window frame/decorations.  see next comment.)*

This approach could skip having to sniff about looking up window IDs and passing them around.

Cool!

P.S.:  I hear what you're saying about debugging xdotool.  I use it in some gvim script I have to do stuff like grow a window across two physical monitors that are united as one logical X11 display.  Most window managers I've encountered consider "full screen" to be "one monitor", so if you want a super wide gvimdiff spanning two displays and want this automated, xdotool is your friend.

**Comment by mmhere** [comment]
Heh.  Way too much code for an enhancement comment, but below ended up doing what I need.

The goal is to reliably pick the top window and move it left/right/up/down a set increment based on which hotkey is used.  A left move followed by a right move should put the window back where it started.  An up move followed by a down move should put the window back where it started.

---

I left all of my debug cruft in there (commented out) so anyone who dares wade into this can see the choices I made.

I'm not much of a Python guy yet, so I cheat by leaning on some ugly old-style shell commands piped to one another to pick up the window frame sizes from xprop.

Note ***in particular*** that these scripts move windows in matched increments accounting for whatever window-frame widths the window manager is using.  So a move-left followed by move-right puts the window right back where it was.  (Same for matched up/down movements.)  This is what the fudge factor stuff is for.

Works on Ubuntu MATE for me;  ***YMMV***.

To wit:

    ###############################################################################
    # Win_Move_Left.py                                                            #
    #     bound to <alt>+<shift>+<left>                                           #
    ###############################################################################
    # Binder script sets values for subsequently called Win_Move.py to use
    xd=-60
    yd=0
    # Surely there is a better way to "pass arguments" ???
    store.set_global_value("xd",xd)
    store.set_global_value("yd",yd)
    engine.run_script("Win_Move")

    ###############################################################################
    # Win_Move_Right.py                                                           #
    #     bound to <alt>+<shift>+<right>                                          #
    ###############################################################################
    xd=60
    yd=0
    store.set_global_value("xd",xd)
    store.set_global_value("yd",yd)
    engine.run_script("Win_Move")

    ###############################################################################
    # Win_Move_Up.py                                                              #
    #     bound to <alt>+<shift>+<up>                                             #
    ###############################################################################
    xd=0
    yd=-65
    store.set_global_value("xd",xd)
    store.set_global_value("yd",yd)
    engine.run_script("Win_Move")

    ###############################################################################
    # Win_Move_Down.py                                                            #
    #     bound to <alt>+<shift>+<down>                                           #
    ###############################################################################
    xd=0
    yd=65
    store.set_global_value("xd",xd)
    store.set_global_value("yd",yd)
    engine.run_script("Win_Move")

    ###############################################################################
    # Win_Move.py                                                                 #
    ###############################################################################
    # Move Active Window By (xd,yd) Deltas--
    #
    # - based on "parameters" wrapper script writes into:
    #   . store.set_global_value("xd",xd)
    #   . store.set_global_value("yd",yd)
    # - surely there is a better way to separate parameterization and core logic!

    # do X11 heavy lifting with sub-shell shenanigans to ensure top window is targeted using its unique ID
    # - the autokey API doesn't quite get us there using merely window.get_active_title() and window.resize_move(title, ...)

    # ---BEGIN---

    # my_date=system.exec_command("~/bin/dt", True) #DBG#
    # dialog.info_dialog("running at datestamp","v 27 -- %s" % (my_date) ) #DBG#

    # identify the top window uniquely:
    # we assume these system.exec_command()s will be in a shell, usually bash:
    # cmd  = 'printf 0x%08x $(xdotool getactivewindow)' # this does work for tools that prefer hex IDs 0x00000000
    # winIDhex = '' # xdotool was unhappy receiving this as hex, so we leave this out
    # winIDhex = system.exec_command(cmd, True)
    cmd  = 'xdotool getactivewindow'
    winID = ''
    winID = system.exec_command(cmd, True)

    # dialog.info_dialog("xdotool getactivewindow returns","cmd '%s'\n winIDhex '%s' winID '%s' use THIS ONE" % (cmd,winIDhex,winID) ) #DBG#

    # The window frame/size info is system wide and should be stored (per machine) but we'll get it each time--
    # - because these autokey script/config files are shared in $HOME across multiple varying machines
    # - each machine's setup (options in window manager or different window manager) may have different values
    # - INEFFICIENT TO REPEAT THIS ALL THE TIME!
    #
    # NB:  couldn't get winIDhex to work with this invocation of xprop for some reason, but he's happy to take decimal winID:
    # - this ugly magic (stripping ',' and then picking positional fields) should be done locally in Python:
    #   Below:
    #   . $3 is the X width of left-hand frame of the window, which we need to fudge the actual X move
    #   . $5 is the Y width of top-side  frame of the window, which we need to fudge the actual Y move
    cmd_base = 'xprop _NET_FRAME_EXTENTS -id ' + winID + ' | sed -e s/,//g | awk '
    cmd1 = cmd_base + "'{print $3}'"
    x_decor = system.exec_command(cmd1, True)
    cmd2 = cmd_base + "'{print $5}'"
    y_decor = system.exec_command(cmd2, True)
    x_fudge = int(x_decor) * 2
    y_fudge = int(y_decor) * 2
    # dialog.info_dialog("xprop pipeline returns","cmd1:  %s\ncmd2:  %s\n    CMD RESULTs:  x_decor '%s', y_decor '%s'\n    FUDGES:  %d,%d" % (cmd1,cmd2,x_decor,y_decor,x_fudge,y_fudge) ) #DBG#

    # (xd,yd) are paramterized by the calling script, one of WinMove_{Left,Right,Up,Down}.py
    #xd=int(store.get_global_value("xd")) # not necessary, their type is retained as int
    #yd=int(store.get_global_value("yd")) # not necessary, their type is retained as int
    xd=store.get_global_value("xd")
    yd=store.get_global_value("yd")
    x=xd-x_fudge
    y=yd-y_fudge
    # EXAMPLE:  xdotool windowmove --relative -- 39846012 100 100
    cmd = "xdotool windowmove --relative -- " + winID + " " + str(x) + " " + str(y)
    # dialog.info_dialog("xdotool move PREFLIGHT","cmd '%s'" % (cmd) ) #DBG#
    # This one returns no stdout:
    system.exec_command(cmd, False)

Actually works quite well for me even when window titles have the same string across more than one window.

**Comment by mmhere** [comment]
It's not really that many commands if you burn off the copious commentary.  Without the double `awk`-use (and instead some basic parsing in Python) the `xprop` pipeline could be invoked once.  Use of `sed` could be avoided too.

    # wrapper script
    xd=-60
    yd=0
    store.set_global_value("xd",xd)
    store.set_global_value("yd",yd)
    engine.run_script("Win_Move")

    #############################################################################
    # Win_Move.py
    cmd  = 'xdotool getactivewindow'
    winID = system.exec_command(cmd, True)

    cmd_base = 'xprop _NET_FRAME_EXTENTS -id ' +winID+ ' | sed -e s/,//g | awk '
    cmd1 = cmd_base + "'{print $3}'"
    x_decor = system.exec_command(cmd1, True)
    cmd2 = cmd_base + "'{print $5}'"
    y_decor = system.exec_command(cmd2, True)
    x_fudge = int(x_decor) * 2
    y_fudge = int(y_decor) * 2

    xd=store.get_global_value("xd")
    yd=store.get_global_value("yd")
    x=xd-x_fudge
    y=yd-y_fudge
    cmd = "xdotool windowmove --relative -- "+winID+ " " + str(x) + " " + str(y)
    system.exec_command(cmd, False)

**Comment by mmhere** [comment]
Heh.  This is arguably simpler (certainly fewer `system.exec_command()` calls, although it does lean pretty heavily on a magic one-liner involving double `xprop` invocation.  That is still a shell pipeline and uses secondary invocations of `cut`.  The final `xdool` command to move active window is simplified.  No more need to actually grab the window ID inside the Python code.

    #############################################################################
    # Win_Move.py
    cmd  = "xprop -id $(xprop -root 0i ',$0' _NET_ACTIVE_WINDOW|cut -d, -f 2) 0i ',$0,$2\n' _NET_FRAME_EXTENTS"
    xprop_out="INIT"
    try:
        xprop_out = system.exec_command(cmd, True)
    except:
        dialog.info_dialog('Win_Move.py:  xprop ERROR',"xprop_out probly not assigned" )
    
    (field,x_decor,y_decor) = xprop_out.split(',')
    x_fudge = int(x_decor) * 2
    y_fudge = int(y_decor) * 2
    
    xd=store.get_global_value("xd")
    yd=store.get_global_value("yd")
    x=xd-x_fudge
    y=yd-y_fudge
    cmd = "xdotool getactivewindow windowmove --relative -- " + str(x) + " " + str(y)
    
    system.exec_command(cmd, False)

**Comment by josephj11** [comment]
Thanks for sharing your code... However, a lot more people would see it if you post such things to our [list](https://groups.google.com/forum/#!forum/autokey-users) rather than as part of an issue.

**Comment by mmhere** [comment]
yeah, i kinda went nutz.  i can delete these here if that would be preferred

**Comment by josephj11** [comment]
No reason to. They're useful examples.

**Comment by sebastiansam55** [comment]
Since I am only seeing this now and don't see it referenced in skimming this issue, you can also use the `wmctrl` native `:SELECT:` function which allows you to manually select the window you want to act upon.

I'm working on expanding the window api currently and would be interested in your feedback on the current implementation as show in #471 

In addition as mentioned in my [comment](https://github.com/autokey/autokey/pull/471#issuecomment-848213193) I am probably going to add a `by_hex` flag to many if not all the methods there, new and old.

Would these changes adequately resolve your issues with the API?
